### PR TITLE
Micro optimization: Use a single resolved promise in AMP

### PR DESCRIPTION
**Pull request by cramforce**
- Eliminates the allocation of the resolved promises
- Makes the function call to get the promise inlinable (as in JS engine)

Benchmark: https://jsperf.com/resolved-promises/1

This might be a really bad idea. Please discuss.

### CODE REVIEW

File: build-system/test-configs/conformance-config.textproto [hunk 1]
```diff
@@ -320,6 +320,15 @@ requirement: {
   whitelist: 'ads/google/deprecated_doubleclick.js'
 }
 
+requirement: {
+  type: RESTRICTED_NAME_CALL
+  error_message: 'import resolvedPromise() instead of calling Promise.resolve() directly.'
+  value: 'Promise.resolve:function(*),!Promise)'
```
Comment (jridgewell) [inline build-system/test-configs/conformance-config.textproto:326]: I think this is supposed to be `Promise.resolve:function(*)`

File: src/resolvedPromise.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+let resolved;
+
+/**
+ * Returns a cached resolved promise.
+ * Use this instead of directly calling Promise.resolve().
+ * @return {!Promise<undefined>}
+ */
+export function resolvedPromise() {
+  if (!resolved) {
+    return (resolved = Promise./* initializer */ resolve());
```
Comment (jridgewell) [inline src/resolvedPromise.js:26]: We should probably freeze this in dev mode.

**Comment by samouri** [comment]
This is cool! 

Would this be better as a pre-Closure Babel transform? That way developers can continue to write `Promise.resolve()` and yet also gain the performance benefit.

**Comment by cramforce** [comment]
Could be done that way as well. Do you want to give it a shot?

Also, for the record, the VM people are saying it is legit
https://twitter.com/cramforce/status/1252740199135670274

**Comment by jridgewell** [comment]
Good idea @samouri. https://github.com/ampproject/amphtml/pull/27960. This'll hit less merge conflicts, too.

**Pull request closed without merging**